File: src/diffler/collectors/v2_output.py
"""Format collected data into stats-action schema v2 output.

This ensures backward compatibility with existing consumers of the
github-user-stats.json format.
"""

from __future__ import annotations

from typing import Any


def build_v2_output(ctx_results: dict[str, Any]) -> dict[str, Any]:
    """Build a schema v2 stats dict from collector results.

    Args:
        ctx_results: The ``results`` dict from a CollectorContext.

    Returns:
        Schema v2-compatible stats JSON.
    """
    profile = ctx_results.get("profile", {})
    contributions = ctx_results.get("contributions", {})
    repos = ctx_results.get("repositories", [])
    gists = ctx_results.get("gists", [])
    repo_stats = ctx_results.get("repo_stats", {})
    computed_stats = ctx_results.get("computed_stats", {})
    activity = ctx_results.get("activity", {})

    # Build top languages from repository data
    lang_bytes: dict[str, dict] = {}
    for repo in repos:
        for lang in repo.get("languages", []):
            name = lang.get("name")
            if not name:
                continue
            if name not in lang_bytes:
                lang_bytes[name] = {
                    "languageName": name,
                    "color": lang.get("color"),
                    "value": 0,
                }
            lang_bytes[name]["value"] += lang.get("size", 0)

    total_bytes = sum(l["value"] for l in lang_bytes.values())
    top_languages = sorted(lang_bytes.values(), key=lambda l: l["value"], reverse=True)
    for lang in top_languages:
        lang["percentage"] = round((lang["value"] / total_bytes) * 100, 2) if total_bytes else 0

    # Build top topics
    topic_counts: dict[str, int] = {}
    for repo in repos:
        for topic in repo.get("topics", []):
            if topic:
                topic_counts[topic] = topic_counts.get(topic, 0) + 1
    top_topics = [
        {"topic": t, "count": c}
        for t, c in sorted(topic_counts.items(), key=lambda x: x[1], reverse=True)
    ][:20]

    public_repos = [r for r in repos if not r.get("is_private")]
    total_stars = sum(r.get("stars", 0) for r in repos)
    total_forks = sum(r.get("forks", 0) for r in repos)

    # Aggregate from backfill collectors if present
    traffic = ctx_results.get("traffic", {})
    contributor_stats = ctx_results.get("contributor_stats", [])
    total_views = sum(v.get("count", 0) for v in traffic.get("views", []))
    total_additions = sum(c.get("additions", 0) for c in contributor_stats)
    total_deletions = sum(c.get("deletions", 0) for c in contributor_stats)

    # Metric cards for remotion
    cards = _build_metric_cards(ctx_results)
    highlights = _build_highlights(ctx_results)

    # Collection status
    collection_status = ctx_results.get("_collection_status", {})

    return {
        "schemaVersion": 2,
        "name": profile.get("name"),
        "username": profile.get("login"),
        "avatarUrl": f"https://avatars.githubusercontent.com/u/{profile.get('login', '')}",
        "bio": profile.get("bio"),
        "company": profile.get("company"),
        "location": profile.get("location"),
        "email": profile.get("email"),
        "twitterUsername": profile.get("twitterUsername"),
        "websiteUrl": profile.get("websiteUrl"),
        "createdAt": profile.get("createdAt"),
        "followers": profile.get("followers", 0),
        "following": profile.get("following", 0),
        "starredRepositories": profile.get("starredRepositories", 0),
        "publicGists": len(gists),
        "totalCommits": contributions.get("totalCommitContributions", 0),
        "totalPullRequests": contributions.get("totalPullRequestContributions", 0),
        "totalPullRequestReviews": contributions.get("totalPullRequestReviewContributions", 0),
        "commitCount": contributions.get("totalCommitContributions", 0),
        "linesOfCodeChanged": total_additions + total_deletions,
        "linesAdded": total_additions,
        "linesDeleted": total_deletions,
        "linesChanged": total_additions + total_deletions,
        "repoViews": total_views,
        "repoViewUniques": sum(v.get("uniques", 0) for v in traffic.get("views", [])),
        "codeByteTotal": total_bytes,
        "topLanguages": top_languages,
        "topTopics": top_topics,
        "starsGiven": activity.get("starsGiven", 0),
        "openIssues": activity.get("openIssues", 0),
        "closedIssues": activity.get("closedIssues", 0),
        "repositoriesContributedTo": activity.get("repositoriesContributedTo", 0),
        "discussionsStarted": activity.get("discussionsStarted", 0),
        "discussionsAnswered": activity.get("discussionsAnswered", 0),
        "profileContributions": {
            "totalContributions": contributions.get("totalContributions", 0),
            "totalCommitContributions": contributions.get("totalCommitContributions", 0),
            "totalIssueContributions": contributions.get("totalIssueContributions", 0),
            "totalPullRequestContributions": contributions.get("totalPullRequestContributions", 0),
            "totalPullRequestReviewContributions": contributions.get("totalPullRequestReviewContributions", 0),
            "totalRepositoryContributions": len(repos),
            "restrictedContributionsCount": contributions.get("restrictedContributionsCount", 0),
            "contributionCalendar": {
                "totalContributions": contributions.get("totalContributions", 0),
                "weeks": contributions.get("calendar", []),
            },
            "stats": {
                "currentStreak": contributions.get("currentStreak", 0),
                "longestStreak": contributions.get("longestStreak", 0),
                "longestStreakStartDate": contributions.get("longestStreakStartDate", ""),
                "longestStreakEndDate": contributions.get("longestStreakEndDate", ""),
            },
        },
        "activity": activity,
        "repositories": [
            {
                "name": r.get("name"),
                "full_name": r.get("full_name"),
                "description": r.get("description"),
                "url": r.get("url"),
                "stars": r.get("stars", 0),
                "forks": r.get("forks", 0),
                "language": r.get("primary_language"),
                "language_color": r.get("primary_language_color"),
                "isPrivate": r.get("is_private", False),
                "isFork": r.get("is_fork", False),
                "isArchived": r.get("is_archived", False),
                "topics": r.get("topics", []),
                "createdAt": r.get("created_at"),
                "pushedAt": r.get("pushed_at"),
                "updatedAt": r.get("updated_at"),
                "owner": r.get("owner"),
                "ownerType": r.get("owner_type"),
                "visibility": r.get("visibility"),
            }
            for r in repos
        ],
        "repoMetrics": {
            "starCount": total_stars,
            "forkCount": total_forks,
            "codeByteTotal": total_bytes,
            "topLanguages": top_languages,
            "topTopics": top_topics,
            "publicRepoCount": len(public_repos),
            "contributorStats": contributor_stats,
            "traffic": traffic,
            "repoStats": repo_stats,
            "computedStats": computed_stats,
        },
        "repoStats": repo_stats,
        "computedStats": computed_stats,
        "cards": cards,
        "highlights": highlights,
        "presentation": {
            "readmeSummary": {
                "contributions": contributions.get("totalContributions", 0),
                "streak": contributions.get("currentStreak", 0),
                "repos": len(repos),
                "stars": total_stars,
            },
            "remotion": {
                "username": profile.get("login"),
                "totalContributions": contributions.get("totalContributions", 0),
                "currentStreak": contributions.get("currentStreak", 0),
                "longestStreak": contributions.get("longestStreak", 0),
                "topLanguages": [
                    {"name": l["languageName"], "percentage": l["percentage"]}
                    for l in top_languages[:6]
                ],
            },
        },
        "privacy": {
            "includePrivateDetails": True,
            "redactedEntries": 0,
        },
        "collectionStatus": {
            "complete": collection_status.get("complete", True),
            "coreComplete": collection_status.get("coreComplete", True),
            "backfillPending": collection_status.get("backfillPending", 0),
            "backfillCompletedThisRun": collection_status.get("backfillCompletedThisRun", 0),
            "backfillFailedThisRun": collection_status.get("backfillFailedThisRun", 0),
            "warnings": collection_status.get("warnings", []),
            "errors": collection_status.get("errors", []),
            "timestamp": collection_status.get("timestamp", ""),
        },
    }


def _build_metric_cards(ctx_results: dict[str, Any]) -> list[dict[str, Any]]:
    """Generate metric cards for remotion consumption."""
    profile = ctx_results.get("profile", {})
    contributions = ctx_results.get("contributions", {})
    repos = ctx_results.get("repositories", [])
    activity = ctx_results.get("activity", {})

    cards: list[dict[str, Any]] = [
        {
            "id": "contributions",
            "label": "Contributions",
            "value": contributions.get("totalContributions", 0),
        },
        {
            "id": "commits",
            "label": "Commits",
            "value": contributions.get("totalCommitContributions", 0),
        },
        {
            "id": "pull-requests",
            "label": "Pull Requests",
            "value": activity.get("totalPullRequests", contributions.get("totalPullRequestContributions", 0)),
        },
        {
            "id": "issues",
            "label": "Issues",
            "value": activity.get("closedIssues", 0),
        },
        {
            "id": "repositories",
            "label": "Repositories",
            "value": len(repos),
        },
        {
            "id": "stars",
            "label": "Stars",
            "value": sum(r.get("stars", 0) for r in repos),
        },
        {
            "id": "followers",
            "label": "Followers",
            "value": profile.get("followers", 0),
        },
    ]

    streak = contributions.get("currentStreak", 0)
    if streak > 0:
        cards.append({
            "id": "streak",
            "label": "Current Streak",
            "value": f"{streak} days",
        })

    return cards


def _build_highlights(ctx_results: dict[str, Any]) -> list[dict[str, Any]]:
    """Generate highlight cards — the most impressive metrics."""
    contributions = ctx_results.get("contributions", {})
    repos = ctx_results.get("repositories", [])
    activity = ctx_results.get("activity", {})
    computed = ctx_results.get("computed_stats", {})

    highlights: list[dict[str, Any]] = []
    total_stars = sum(r.get("stars", 0) for r in repos)

    if contributions.get("currentStreak", 0) >= 7:
        highlights.append({
            "id": "streak",
            "label": "🔥 Streak",
            "value": f"{contributions['currentStreak']} days",
            "detail": f"Longest: {contributions.get('longestStreak', 0)} days",
        })

    if total_stars >= 50:
        highlights.append({
            "id": "popular",
            "label": "⭐ Popular",
            "value": f"{total_stars} stars",
            "detail": f"Across {len(repos)} repos",
        })

    if activity.get("repositoriesContributedTo", 0) >= 5:
        highlights.append({
            "id": "contributor",
            "label": "🌐 Contributor",
            "value": f"{activity['repositoriesContributedTo']} repos",
        })

    yoy = computed.get("yearOverYearGrowth", 0)
    if yoy > 0:
        highlights.append({
            "id": "growth",
            "label": "📈 Growth",
            "value": f"+{yoy}%",
            "detail": "Year over year",
        })

    return highlights

File: src/collectors/v2Output.ts
interface LangBytes {
  languageName: string;
  color: string | null;
  value: number;
  percentage?: number;
}

interface RepoData {
  name?: string;
  full_name?: string;
  description?: string | null;
  url?: string;
  stars?: number;
  forks?: number;
  primary_language?: string | null;
  primary_language_color?: string | null;
  is_private?: boolean;
  is_fork?: boolean;
  is_archived?: boolean;
  topics?: string[];
  created_at?: string | null;
  pushed_at?: string | null;
  updated_at?: string | null;
  owner?: string | null;
  owner_type?: string | null;
  visibility?: string | null;
  languages?: Array<{ name: string; color: string | null; size: number }>;
}

function buildMetricCards(ctxResults: Record<string, unknown>): Array<Record<string, unknown>> {
  const profile = (ctxResults.profile as Record<string, unknown>) ?? {};
  const contributions = (ctxResults.contributions as Record<string, unknown>) ?? {};
  const repos = (ctxResults.repositories as RepoData[]) ?? [];
  const activity = (ctxResults.activity as Record<string, number>) ?? {};

  const cards: Array<Record<string, unknown>> = [
    { id: "contributions", label: "Contributions", value: contributions.totalContributions ?? 0 },
    { id: "commits", label: "Commits", value: contributions.totalCommitContributions ?? 0 },
    {
      id: "pull-requests",
      label: "Pull Requests",
      value: activity.totalPullRequests ?? contributions.totalPullRequestContributions ?? 0,
    },
    { id: "issues", label: "Issues", value: activity.closedIssues ?? 0 },
    { id: "repositories", label: "Repositories", value: repos.length },
    { id: "stars", label: "Stars", value: repos.reduce((a, r) => a + (r.stars ?? 0), 0) },
    { id: "followers", label: "Followers", value: profile.followers ?? 0 },
  ];

  const streak = (contributions.currentStreak as number) ?? 0;
  if (streak > 0) {
    cards.push({ id: "streak", label: "Current Streak", value: `${streak} days` });
  }

  return cards;
}

function buildHighlights(ctxResults: Record<string, unknown>): Array<Record<string, unknown>> {
  const contributions = (ctxResults.contributions as Record<string, unknown>) ?? {};
  const repos = (ctxResults.repositories as RepoData[]) ?? [];
  const activity = (ctxResults.activity as Record<string, number>) ?? {};
  const computed = (ctxResults.computed_stats as Record<string, unknown>) ?? {};

  const highlights: Array<Record<string, unknown>> = [];
  const totalStars = repos.reduce((a, r) => a + (r.stars ?? 0), 0);

  if ((contributions.currentStreak as number) >= 7) {
    highlights.push({
      id: "streak",
      label: "🔥 Streak",
      value: `${contributions.currentStreak} days`,
      detail: `Longest: ${contributions.longestStreak ?? 0} days`,
    });
  }

  if (totalStars >= 50) {
    highlights.push({
      id: "popular",
      label: "⭐ Popular",
      value: `${totalStars} stars`,
      detail: `Across ${repos.length} repos`,
    });
  }

  if ((activity.repositoriesContributedTo ?? 0) >= 5) {
    highlights.push({
      id: "contributor",
      label: "🌐 Contributor",
      value: `${activity.repositoriesContributedTo} repos`,
    });
  }

  const yoy = (computed.yearOverYearGrowth as number) ?? 0;
  if (yoy > 0) {
    highlights.push({
      id: "growth",
      label: "📈 Growth",
      value: `+${yoy}%`,
      detail: "Year over year",
    });
  }

  return highlights;
}

export function buildV2Output(ctxResults: Record<string, unknown>): Record<string, unknown> {
  const profile = (ctxResults.profile as Record<string, unknown>) ?? {};
  const contributions = (ctxResults.contributions as Record<string, unknown>) ?? {};
  const repos = (ctxResults.repositories as RepoData[]) ?? [];
  const gists = (ctxResults.gists as Array<Record<string, unknown>>) ?? [];
  const repoStats = (ctxResults.repo_stats as Record<string, unknown>) ?? {};
  const computedStats = (ctxResults.computed_stats as Record<string, unknown>) ?? {};
  const activity = (ctxResults.activity as Record<string, number>) ?? {};
  const traffic = (ctxResults.traffic as Record<string, Array<Record<string, unknown>>>) ?? {};
  const contributorStats = (ctxResults.contributor_stats as Array<Record<string, unknown>>) ?? [];
  const collectionStatus = (ctxResults._collection_status as Record<string, unknown>) ?? {};

  // Build top languages
  const langBytes: Record<string, LangBytes> = {};
  for (const repo of repos) {
    for (const lang of repo.languages ?? []) {
      const name = lang.name;
      if (!name) continue;
      if (!langBytes[name]) {
        langBytes[name] = { languageName: name, color: lang.color ?? null, value: 0 };
      }
      langBytes[name].value += lang.size;
    }
  }

  const totalBytes = Object.values(langBytes).reduce((a, l) => a + l.value, 0);
  const topLanguages = Object.values(langBytes)
    .sort((a, b) => b.value - a.value)
    .map((l) => ({
      ...l,
      percentage: totalBytes > 0 ? Math.round((l.value / totalBytes) * 10000) / 100 : 0,
    }));

  // Build top topics
  const topicCounts: Record<string, number> = {};
  for (const repo of repos) {
    for (const topic of repo.topics ?? []) {
      if (topic) {
        topicCounts[topic] = (topicCounts[topic] ?? 0) + 1;
      }
    }
  }
  const topTopics = Object.entries(topicCounts)
    .sort((a, b) => b[1] - a[1])
    .slice(0, 20)
    .map(([topic, count]) => ({ topic, count }));

  const publicRepos = repos.filter((r) => !r.is_private);
  const totalStars = repos.reduce((a, r) => a + (r.stars ?? 0), 0);
  const totalForks = repos.reduce((a, r) => a + (r.forks ?? 0), 0);

  const totalViews = (traffic.views ?? []).reduce((a, v) => a + ((v.count as number) ?? 0), 0);
  const totalAdditions = contributorStats.reduce((a, c) => a + ((c.additions as number) ?? 0), 0);
  const totalDeletions = contributorStats.reduce((a, c) => a + ((c.deletions as number) ?? 0), 0);

  const cards = buildMetricCards(ctxResults);
  const highlights = buildHighlights(ctxResults);

  return {
    schemaVersion: 2,
    name: profile.name,
    username: profile.login,
    avatarUrl: `https://avatars.githubusercontent.com/u/${profile.login ?? ""}`,
    bio: profile.bio,
    company: profile.company,
    location: profile.location,
    email: profile.email,
    twitterUsername: profile.twitterUsername,
    websiteUrl: profile.websiteUrl,
    createdAt: profile.createdAt,
    followers: profile.followers ?? 0,
    following: profile.following ?? 0,
    starredRepositories: profile.starredRepositories ?? 0,
    publicGists: gists.length,
    totalCommits: contributions.totalCommitContributions ?? 0,
    totalPullRequests: contributions.totalPullRequestContributions ?? 0,
    totalPullRequestReviews: contributions.totalPullRequestReviewContributions ?? 0,
    commitCount: contributions.totalCommitContributions ?? 0,
    linesOfCodeChanged: totalAdditions + totalDeletions,
    linesAdded: totalAdditions,
    linesDeleted: totalDeletions,
    linesChanged: totalAdditions + totalDeletions,
    repoViews: totalViews,
    repoViewUniques: (traffic.views ?? []).reduce((a, v) => a + ((v.uniques as number) ?? 0), 0),
    codeByteTotal: totalBytes,
    topLanguages,
    topTopics,
    starsGiven: activity.starsGiven ?? 0,
    openIssues: activity.openIssues ?? 0,
    closedIssues: activity.closedIssues ?? 0,
    repositoriesContributedTo: activity.repositoriesContributedTo ?? 0,
    discussionsStarted: activity.discussionsStarted ?? 0,
    discussionsAnswered: activity.discussionsAnswered ?? 0,
    profileContributions: {
      totalContributions: contributions.totalContributions ?? 0,
      totalCommitContributions: contributions.totalCommitContributions ?? 0,
      totalIssueContributions: contributions.totalIssueContributions ?? 0,
      totalPullRequestContributions: contributions.totalPullRequestContributions ?? 0,
      totalPullRequestReviewContributions: contributions.totalPullRequestReviewContributions ?? 0,
      totalRepositoryContributions: repos.length,
      restrictedContributionsCount: contributions.restrictedContributionsCount ?? 0,
      contributionCalendar: {
        totalContributions: contributions.totalContributions ?? 0,
        weeks: contributions.calendar ?? [],
      },
      stats: {
        currentStreak: contributions.currentStreak ?? 0,
        longestStreak: contributions.longestStreak ?? 0,
        longestStreakStartDate: contributions.longestStreakStartDate ?? "",
        longestStreakEndDate: contributions.longestStreakEndDate ?? "",
      },
    },
    activity,
    repositories: repos.map((r) => ({
      name: r.name,
      full_name: r.full_name,
      description: r.description,
      url: r.url,
      stars: r.stars ?? 0,
      forks: r.forks ?? 0,
      language: r.primary_language,
      language_color: r.primary_language_color,
      isPrivate: r.is_private ?? false,
      isFork: r.is_fork ?? false,
      isArchived: r.is_archived ?? false,
      topics: r.topics ?? [],
      createdAt: r.created_at,
      pushedAt: r.pushed_at,
      updatedAt: r.updated_at,
      owner: r.owner,
      ownerType: r.owner_type,
      visibility: r.visibility,
    })),
    repoMetrics: {
      starCount: totalStars,
      forkCount: totalForks,
      codeByteTotal: totalBytes,
      topLanguages,
      topTopics,
      publicRepoCount: publicRepos.length,
      contributorStats: contributorStats,
      traffic,
      repoStats,
      computedStats,
    },
    repoStats,
    computedStats,
    cards,
    highlights,
    presentation: {
      readmeSummary: {
        contributions: contributions.totalContributions ?? 0,
        streak: contributions.currentStreak ?? 0,
        repos: repos.length,
        stars: totalStars,
      },
      remotion: {
        username: profile.login,
        totalContributions: contributions.totalContributions ?? 0,
        currentStreak: contributions.currentStreak ?? 0,
        longestStreak: contributions.longestStreak ?? 0,
        topLanguages: topLanguages.slice(0, 6).map((l) => ({
          name: l.languageName,
          percentage: l.percentage,
        })),
      },
    },
    privacy: {
      includePrivateDetails: true,
      redactedEntries: 0,
    },
    collectionStatus: {
      complete: collectionStatus.complete ?? true,
      coreComplete: collectionStatus.coreComplete ?? true,
      backfillPending: collectionStatus.backfillPending ?? 0,
      backfillCompletedThisRun: collectionStatus.backfillCompletedThisRun ?? 0,
      backfillFailedThisRun: collectionStatus.backfillFailedThisRun ?? 0,
      warnings: collectionStatus.warnings ?? [],
      errors: collectionStatus.errors ?? [],
      timestamp: collectionStatus.timestamp ?? "",
    },
  };
}
